src/NHibernate/Persister/Entity/AbstractEntityPersister.cs

```diff
@@ -2516,21 +2516,29 @@ protected IUniqueEntityLoader CreateEntityLoader(LockMode lockMode)
 			return CreateEntityLoader(lockMode, CollectionHelper.EmptyDictionary<string, IFilter>());
 		}
 
+		//TODO 6.0: Remove (replaced by overload with optional parameter) 
 		protected bool Check(int rows, object id, int tableNumber, IExpectation expectation, DbCommand statement)
+		{
+			return Check(rows, id, tableNumber, expectation, statement, false);
+		}
+
+		protected bool Check(int rows, object id, int tableNumber, IExpectation expectation, DbCommand statement, bool forceThrowStaleException = false)
 		{
 			try
 			{
 				expectation.VerifyOutcomeNonBatched(rows, statement);
 			}
 			catch (StaleStateException sse)
 			{
-				if (!IsNullableTable(tableNumber))
+				if (forceThrowStaleException || !IsNullableTable(tableNumber))
 				{
 					if (Factory.Statistics.IsStatisticsEnabled)
 						Factory.StatisticsImplementor.OptimisticFailure(EntityName);
 
 					throw new StaleObjectStateException(EntityName, id, sse);
 				}
+
+				return false;
```

Comment:
Previously this were returning `true`. Why change?

Reply:
Needed to properly support saving detached entity. I believe this change is covered by PR tests.

Reply:
Test is `UpdateNullOptionalJoinToNotNullDetached`